File: backend/app/schemas/challenge.py
"""
Challenge schemas for request/response validation.
"""

from pydantic import Field, field_validator, model_validator
from typing import List, Optional
from uuid import UUID
from datetime import datetime

from app.schemas.base import BaseSchema, PaginatedResponse
from app.models.enums import ChallengeStatus, PriorityLevel, MediaType


class CategoryResponse(BaseSchema):
    """Schema for category response."""
    
    id: UUID
    name: str
    description: Optional[str] = None
    parent_id: Optional[UUID] = None


class ChallengeCreate(BaseSchema):
    """Schema for creating a challenge."""
    
    title: str = Field(..., min_length=10, max_length=500, description="Challenge title")
    description: str = Field(..., min_length=30, max_length=5000, description="Detailed description")
    district: str = Field(..., min_length=2, max_length=100, description="District name")
    block: Optional[str] = Field(None, max_length=100, description="Block name")
    village: Optional[str] = Field(None, max_length=100, description="Village name")
    latitude: Optional[float] = Field(None, ge=-90, le=90, description="GPS latitude")
    longitude: Optional[float] = Field(None, ge=-180, le=180, description="GPS longitude")
    affected_population: Optional[int] = Field(None, ge=0, le=10000000, description="Estimated affected population")
    category_ids: List[UUID] = Field(..., min_length=1, description="At least one category ID")
    
    @field_validator('title', 'description', 'district', 'block', 'village')
    @classmethod
    def strip_whitespace(cls, v):
        """Strip whitespace and reject blank strings."""
        if v is None:
            return v
        
        stripped = v.strip()
        if not stripped:
            raise ValueError('Field cannot be blank or whitespace only')
        
        return stripped
    
    @model_validator(mode='after')
    def validate_coordinates(self):
        """Validate that both coordinates are provided together."""
        lat = self.latitude
        lon = self.longitude
        
        # If one is provided, both must be provided
        if (lat is not None and lon is None) or (lat is None and lon is not None):
            raise ValueError('Both latitude and longitude must be provided together')
        
        return self


class ChallengeUpdate(BaseSchema):
    """Schema for updating a challenge."""
    
    title: Optional[str] = Field(None, min_length=10, max_length=500)
    description: Optional[str] = Field(None, min_length=30, max_length=5000)
    district: Optional[str] = Field(None, min_length=2, max_length=100)
    block: Optional[str] = Field(None, max_length=100)
    village: Optional[str] = Field(None, max_length=100)
    latitude: Optional[float] = Field(None, ge=-90, le=90)
    longitude: Optional[float] = Field(None, ge=-180, le=180)
    affected_population: Optional[int] = Field(None, ge=0, le=10000000)
    category_ids: Optional[List[UUID]] = Field(None, min_length=1)
    
    @field_validator('title', 'description', 'district', 'block', 'village')
    @classmethod
    def strip_whitespace(cls, v):
        """Strip whitespace and reject blank strings."""
        if v is None:
            return v
        
        stripped = v.strip()
        if not stripped:
            raise ValueError('Field cannot be blank or whitespace only')
        
        return stripped


class ChallengeMediaResponse(BaseSchema):
    """Schema for media file response."""
    
    id: UUID
    media_type: MediaType
    file_url: str
    file_name: str
    mime_type: Optional[str]
    file_size: Optional[int]
    created_at: datetime


class ChallengeResponse(BaseSchema):
    """Schema for challenge response."""
    
    id: UUID
    challenge_code: str
    title: str
    description: str
    status: ChallengeStatus
    priority_score: Optional[float] = None
    priority_level: Optional[PriorityLevel] = None
    district: str
    block: Optional[str] = None
    village: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    affected_population: Optional[int] = None
    categories: List[CategoryResponse] = []
    media: List[ChallengeMediaResponse] = []
    created_at: datetime
    updated_at: datetime
    
    # Submitter info (optional, depends on access level)
    submitted_by: Optional[UUID] = None
    submitter_name: Optional[str] = None


class ChallengeListItem(BaseSchema):
    """Schema for challenge in list view (lighter than full response)."""
    
    id: UUID
    challenge_code: str
    title: str
    status: ChallengeStatus
    priority_level: Optional[PriorityLevel] = None
    district: str
    affected_population: Optional[int] = None
    category_count: int = 0
    media_count: int = 0
    created_at: datetime


class ChallengeListResponse(BaseSchema):
    """Schema for paginated challenge list response."""
    
    items: List[ChallengeListItem]
    page: int
    page_size: int
    total: int
    pages: int
    has_next: bool
    has_previous: bool


class ChallengeFilterParams(BaseSchema):
    """Schema for challenge filtering parameters."""
    
    status: Optional[ChallengeStatus] = None
    district: Optional[str] = None
    category_id: Optional[UUID] = None
    priority_level: Optional[PriorityLevel] = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)

File: frontend/lib/validation/challenge.ts
/**
 * Challenge validation utilities
 * Matches backend validation constraints exactly
 */

export interface ChallengeValidationErrors {
  title?: string
  description?: string
  district?: string
  block?: string
  village?: string
  latitude?: string
  longitude?: string
  affected_population?: string
  category_ids?: string
}

/**
 * Validate challenge title
 * Backend: min 10, max 500 characters, cannot be blank/whitespace
 */
export function validateTitle(title: string): string | undefined {
  const trimmed = title.trim()
  
  if (!trimmed) {
    return 'Title is required'
  }
  
  if (trimmed.length < 10) {
    return 'Title must be at least 10 characters'
  }
  
  if (trimmed.length > 500) {
    return 'Title must not exceed 500 characters'
  }
  
  return undefined
}

/**
 * Validate challenge description
 * Backend: min 30, max 5000 characters, cannot be blank/whitespace
 */
export function validateDescription(description: string): string | undefined {
  const trimmed = description.trim()
  
  if (!trimmed) {
    return 'Description is required'
  }
  
  if (trimmed.length < 30) {
    return 'Description must be at least 30 characters'
  }
  
  if (trimmed.length > 5000) {
    return 'Description must not exceed 5000 characters'
  }
  
  return undefined
}

/**
 * Validate district name
 * Backend: min 2, max 100 characters, cannot be blank/whitespace
 */
export function validateDistrict(district: string): string | undefined {
  const trimmed = district.trim()
  
  if (!trimmed) {
    return 'District is required'
  }
  
  if (trimmed.length < 2) {
    return 'District must be at least 2 characters'
  }
  
  if (trimmed.length > 100) {
    return 'District must not exceed 100 characters'
  }
  
  return undefined
}

/**
 * Validate block name (optional)
 * Backend: max 100 characters, cannot be whitespace-only
 */
export function validateBlock(block: string | undefined): string | undefined {
  if (!block) return undefined
  
  const trimmed = block.trim()
  if (!trimmed) {
    return 'Block cannot be whitespace only'
  }
  
  if (trimmed.length > 100) {
    return 'Block must not exceed 100 characters'
  }
  
  return undefined
}

/**
 * Validate village name (optional)
 * Backend: max 100 characters, cannot be whitespace-only
 */
export function validateVillage(village: string | undefined): string | undefined {
  if (!village) return undefined
  
  const trimmed = village.trim()
  if (!trimmed) {
    return 'Village cannot be whitespace only'
  }
  
  if (trimmed.length > 100) {
    return 'Village must not exceed 100 characters'
  }
  
  return undefined
}

/**
 * Validate GPS coordinates
 * Backend: latitude -90 to 90, longitude -180 to 180
 * Both must be provided together
 */
export function validateCoordinates(
  latitude: number | null,
  longitude: number | null
): { latitude?: string; longitude?: string } {
  const errors: { latitude?: string; longitude?: string } = {}
  
  // If one is provided, both must be provided
  if (latitude !== null && longitude === null) {
    errors.longitude = 'Longitude is required when latitude is provided'
  }
  
  if (longitude !== null && latitude === null) {
    errors.latitude = 'Latitude is required when longitude is provided'
  }
  
  // Validate ranges
  if (latitude !== null) {
    if (latitude < -90 || latitude > 90) {
      errors.latitude = 'Latitude must be between -90 and 90'
    }
  }
  
  if (longitude !== null) {
    if (longitude < -180 || longitude > 180) {
      errors.longitude = 'Longitude must be between -180 and 180'
    }
  }
  
  return errors
}

/**
 * Validate affected population (optional)
 * Backend: 0 to 10,000,000
 */
export function validateAffectedPopulation(
  population: number | undefined
): string | undefined {
  if (population === undefined || population === null) return undefined
  
  if (population < 0) {
    return 'Affected population cannot be negative'
  }
  
  if (population > 10000000) {
    return 'Affected population must not exceed 10,000,000'
  }
  
  return undefined
}

/**
 * Validate category selection
 * Backend: at least one category required
 */
export function validateCategories(categoryIds: string[]): string | undefined {
  if (!categoryIds || categoryIds.length === 0) {
    return 'At least one category is required'
  }
  
  return undefined
}

/**
 * Validate entire challenge form
 * Returns all validation errors
 */
export function validateChallengeForm(data: {
  title: string
  description: string
  district: string
  block?: string
  village?: string
  latitude?: number | null
  longitude?: number | null
  affected_population?: number
  category_ids: string[]
}): ChallengeValidationErrors {
  const errors: ChallengeValidationErrors = {}
  
  const titleError = validateTitle(data.title)
  if (titleError) errors.title = titleError
  
  const descriptionError = validateDescription(data.description)
  if (descriptionError) errors.description = descriptionError
  
  const districtError = validateDistrict(data.district)
  if (districtError) errors.district = districtError
  
  const blockError = validateBlock(data.block)
  if (blockError) errors.block = blockError
  
  const villageError = validateVillage(data.village)
  if (villageError) errors.village = villageError
  
  const coordErrors = validateCoordinates(
    data.latitude ?? null,
    data.longitude ?? null
  )
  if (coordErrors.latitude) errors.latitude = coordErrors.latitude
  if (coordErrors.longitude) errors.longitude = coordErrors.longitude
  
  const populationError = validateAffectedPopulation(data.affected_population)
  if (populationError) errors.affected_population = populationError
  
  const categoryError = validateCategories(data.category_ids)
  if (categoryError) errors.category_ids = categoryError
  
  return errors
}

/**
 * Check if there are any validation errors
 */
export function hasValidationErrors(errors: ChallengeValidationErrors): boolean {
  return Object.keys(errors).length > 0
}
